Allow multiple azure instances

noaccOS opened the pull request:
Allow passing connection parameters together with the container to blob functions.

Values from the parameters are prioritized over default configuration.

Fixes #44

crertel-getthru commented:
This is super neat! @jakobht can you run the workflow?

jakobht commented on lib/azurex/blob.ex:
```diff
@@ -54,6 +57,9 @@ defmodule Azurex.Blob do
       :ok
+      
+      iex> put_blob("filename.txt", "file contents", "text/plain", [container: "container"])
+      :ok
```
Could you add a more comprehensive example here - like with all the overwrites availible?

jakobht commented on lib/azurex/blob/config.ex:
```diff
@@ -93,10 +100,25 @@ defmodule Azurex.Blob.Config do
+  Returns the given configuration keyword, and defaults to setting the container if given a string.
+  """
+  @spec get_connection_params(config_overrides) :: keyword()
+  def get_connection_params(container) when is_binary(container), do: [container: container]
```
Maybe add a comment here saying this is for backwards compatability

jakobht commented:
Hi, added a few comments, but generally looks good! The CI failed, the linter is sad.

I just updated the build pipelines in master to check 1.16 (but drop 1.11), so please merge. 

Thanks a lot for the PR!

noaccOS commented:
Rebased and added some more examples. I've also mentioned this feature briefly in the README.

The linter is sad because of unused optional parameters but I haven't touched that section myself.
Maybe I could remove those optional parameters for now? It's always possible to add them back later if we need them as it's a private function anyways

stilist commented on lib/azurex/blob.ex:
```diff
@@ -170,21 +189,22 @@ defmodule Azurex.Blob do
+  @spec copy_blob(String.t(), String.t(), Config.config_overrides()) ::
           {:ok, HTTPoison.Response.t()} | {:error, term()}
-  def copy_blob(source_name, destination_name, container \\ nil) do
+  def copy_blob(source_name, destination_name, overrides \\ []) do
```
I suggest documenting that the overrides apply to both source and destination, so you can't use this method to copy between accounts.

jakobht replied:
Not super important IMO, but documentation is always good.

noaccOS replied:
I have added a line in the doc mentioning it

jakobht commented:
> Rebased and added some more examples. I've also mentioned this feature briefly in the README.
> 
> The linter is sad because of unused optional parameters but I haven't touched that section myself. Maybe I could remove those optional parameters for now? It's always possible to add them back later if we need them as it's a private function anyways

It should be fine to delete the unused args. 

Sorry I'm slow at answering :)

stilist commented:
This PR has been very helpful for a project I'm working on — thanks!

I did notice that `SharedAccessSignature.sas_url/3` hasn't been modified to accept config overrides.

noaccOS commented:
> I did notice that `SharedAccessSignature.sas_url/3` hasn't been modified to accept config overrides.

I implemented the same logic for the above function, but in doing so I have made the container optional (ie it defaults to `Config.default_container` like everywhere else)

mattpolzin commented:
@jakobht bumping this in case it fell through the cracks. no worries if you've just been too busy to take another look.

jakobht commented:
Hi, sorry I never got back to this, it's merged now
